crates/turborepo-lib/src/task_graph/visitor.rs

```diff
@@ -838,21 +860,49 @@ impl ExecContext {
 
         match exit_status {
             ChildExit::Finished(Some(0)) => {
+                // Attempt to flush stdout_writer and log any errors encountered
                 if let Err(e) = stdout_writer.flush() {
                     error!("{e}");
-                } else if let Err(e) = self
-                    .task_cache
-                    .save_outputs(&mut prefixed_ui, task_duration, telemetry)
-                    .await
-                {
-                    error!("error caching output: {e}");
                 } else {
-                    self.hash_tracker.insert_expanded_outputs(
-                        self.task_id.clone(),
-                        self.task_cache.expanded_outputs().to_vec(),
-                    );
+                    let mut cacheable = true;
+                    // If stdout_writer flush is successful, proceed with task processing
+                    if self.task_access.is_enabled() {
+                        // Task access enabled: Handle trace file processing
+                        if let Some(trace) =
+                            TaskAccessTraceFile::read(&self.task_access.repo_root, &self.task_hash)
+                        {
+                            // Trace file found
+                            if trace.can_cache(&self.task_access.repo_root) {
+                                // Cache trace and save task outputs if trace is cacheable
+                                self.task_access
+                                    .save_trace(trace, self.task_id_for_display.to_string());
+                            } else {
+                                cacheable = false;
+                            }
+                        }
+                    } else {
+                        // Task access not enabled: Directly save task outputs
+                    }
+
+                    if cacheable {
+                        // Attempt to save task outputs and log any errors encountered
+                        if let Err(e) = self
+                            .task_cache
+                            .save_outputs(&mut prefixed_ui, task_duration, telemetry)
+                            .await
+                        {
+                            error!("error caching output: {e}");
+                        } else {
+                            // If no errors, update hash tracker with expanded outputs
+                            self.hash_tracker.insert_expanded_outputs(
+                                self.task_id.clone(),
+                                self.task_cache.expanded_outputs().to_vec(),
+                            );
+                        }
+                    }
```

Comment:
This "can we cache" logic feels like it should belong totally in the task access module:
```
    // Whether we can cache the given task, returning None if task access isn't
    // enabled or the trace can't be found
    pub fn can_cache(&self, task_hash: &str, task_id: &str) -> Option<bool> {
        if !self.is_enabled() {
            return None;
        }
        let trace = TaskAccessTraceFile::read(&self.repo_root, task_hash)?;
        if trace.can_cache(&self.repo_root) {
            self.save_trace(trace, task_id.to_string());
            Some(true)
        } else {
            Some(false)
        }
    }
```
Which simplifies the changes in Visitor to
```suggestion
                   if self
                        .task_access
                        .can_cache(&self.task_hash, &self.task_id_for_display)
                        .unwrap_or(true)
                    {
                        if let Err(e) = self
                            .task_cache
                            .save_outputs(&mut prefixed_ui, task_duration, telemetry)
                            .await
                        {
                            error!("error caching output: {e}");
                        } else {
                            // If no errors, update hash tracker with expanded outputs
                            self.hash_tracker.insert_expanded_outputs(
                                self.task_id.clone(),
                                self.task_cache.expanded_outputs().to_vec(),
                            );
                        }
                    }
```